Sources/AsyncAlgorithms/AsyncChannel.swift

```diff
@@ -50,68 +53,49 @@ public final class AsyncChannel<Element: Sendable>: AsyncSequence, Sendable {
     }
   }
   
-  struct Awaiting: Hashable {
+  typealias Pending = ChannelToken<UnsafeContinuation<UnsafeContinuation<Element?, Never>?, Never>>
+  typealias Awaiting = ChannelToken<UnsafeContinuation<Element?, Never>>
+
+  struct ChannelToken<Continuation>: Hashable {
     var generation: Int
-    var continuation: UnsafeContinuation<Element?, Never>?
-    let cancelled: Bool
-    
-    init(generation: Int, continuation: UnsafeContinuation<Element?, Never>) {
+    var continuation: Continuation?
+
+    init(generation: Int, continuation: Continuation) {
       self.generation = generation
       self.continuation = continuation
-      cancelled = false
     }
-    
+
     init(placeholder generation: Int) {
       self.generation = generation
       self.continuation = nil
-      cancelled = false
     }
-    
-    init(cancelled generation: Int) {
-      self.generation = generation
-      self.continuation = nil
-      cancelled = true
-    }
-    
+
     func hash(into hasher: inout Hasher) {
       hasher.combine(generation)
     }
-    
-    static func == (_ lhs: Awaiting, _ rhs: Awaiting) -> Bool {
+
+    static func == (_ lhs: ChannelToken, _ rhs: ChannelToken) -> Bool {
       return lhs.generation == rhs.generation
     }
   }
+
+  enum ChannelTokenStatus: Equatable {
+    case new
+    case cancelled
+  }
   
   enum Emission {
     case idle
-    case pending([UnsafeContinuation<UnsafeContinuation<Element?, Never>?, Never>])
+    case pending(Set<Pending>)
```

Comment:
hmm making this a set becomes non-deterministic on the order of application - which will make it hard to test. Do we need set uniqueness to prevent issues or would just an Array work?

Reply:
Unfortunately we have to be able to uniquely identify the pending operations so we can remove/resume them in case of cancellation thanks to their "id".

Using an array as a FIFO could be nice to guarantee deterministic operation but we cannot rely on the index of a newly added operation to potentially remove it in case of a cancellation because the array could be updated by another concurrent send/next operation in the meantime. Also, if want want to keep an O(1) complexity for retrieving a pending operation, we need an indexed storage of some sort.

I guess we need something like OrderedSet but it would imply a dependency on swift-collections which we don't want I presume. Do you think we should implement a custom storage that fits our needs here ?

Reply:
@phausler i see in this PR https://github.com/apple/swift-async-algorithms/pull/185 that an dependency on Swift collections is introduced. Is this something we want ? We could leverage OrderedSet if so

Reply:
@phausler did you had the change to think a bit about having an external dep ?

Reply:
OrderedSet is perfectly fine in my book.

Reply:
Nice, then I’ll update this PR to integrate it.
Thanks for your reply.

Reply:
@phausler I have successfully introduced OrderedSet in this PR

Reply:
@phausler do you think we have a go for this one, now that we use OrderedSet ?

Reply:
yea I think we are good to go here